### PR TITLE
gnrc: enhance documentation

### PR DESCRIPTION
**Pull request by cgundogan**
This PR includes a preliminary version of gnrc's documentation. However, much more info can be added to that. @authmillenon can you add more info where you think it's necessary?

### CODE REVIEW

File: sys/net/gnrc/doc.txt [hunk 1]
```diff
@@ -0,0 +1,169 @@
+/**
+ * @defgroup    net_gnrc Generic (gnrc) network stack.
+ * @ingroup     net
+ * @brief       RIOT's modular default IP network stack.
+ *
+ * @section gnrc_sec_about About
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: Use markdown where you can instead of doxygen tags

**Comment by cgundogan** [comment]
addressed @authmillenon's comment

File: sys/net/gnrc/doc.txt [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+ * @defgroup    net_gnrc Generic (gnrc) network stack.
+ * @ingroup     net
+ * @brief       RIOT's modular default IP network stack.
+ *
+ * # About
+ *
+ * This module is currently the default network stack of RIOT and includes many components
+ * ranging from a simple mac protocol (@ref net_gnrc_nomac) through a fully-featured IPv6
+ * implementation (@ref net_gnrc_ipv6) with 6LowPAN extensions (@ref net_gnrc_sixlowpan) to
+ * an udp implementation (@ref net_gnrc_udp) and the Routing Protocol for Low Power and Lossy
```
Comment (OlegHahm) [inline sys/net/gnrc/doc.txt]: `s/ udp / UDP ` and why is RPL the only thing that is not abbreviated here?

File: sys/net/gnrc/doc.txt [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+ * Networks (@ref net_gnrc_rpl).
+ *
+ * A list of all features contained in the @ref net_gnrc is available in the `Modules` section above.
+ *
+ * # Integration into RIOT-OS
+ *
+ * The integration of the @ref net_gnrc into RIOT-OS is handled by two modules:
```
Comment (OlegHahm) [inline sys/net/gnrc/doc.txt]: `s/RIOT-OS/RIOT`

Comment (OlegHahm) [inline sys/net/gnrc/doc.txt]: Also this sounds somehow odd. Can you rephrase? Something like `from the application layer GNRC can be accessed through the conn interface, while its interface to the network driver is defined by netdev2` or so.

File: sys/net/gnrc/doc.txt (hunk 2 above)
Comment (cgundogan) [inline sys/net/gnrc/doc.txt]: I am biased (: good hint - will update (most likely later in the morning)

File: sys/net/gnrc/doc.txt [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+ * - @ref net_gnrc_netdev2
+ *
+ * The @ref net_gnrc_conn implements the API to the application layer (@ref net_conn),
+ * while the @ref net_gnrc_netdev2 implements the API to communicate with low-level
+ * network device drivers (@ref drivers_netdev_netdev2).
+ *
+ * # Architecture
+ *
+ * ![GNRC Network Stack](riot-gnrc.svg)
+ * Each layer of the network stack resembles a stand-alone thread that runs with a thread priority
```
Comment (OlegHahm) [inline sys/net/gnrc/doc.txt]: Proposal:
"Each layer of the network stack runs in its own thread. Each lower layer thread runs with a higher priority than any upper layer thread.`

File: sys/net/gnrc/doc.txt [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+ * which is decreasing from bottom to top. In this regard, the thread of the mac layer
```
Comment (OlegHahm) [inline sys/net/gnrc/doc.txt]: `s/mac/MAC`

File: sys/net/gnrc/doc.txt [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+ * implementation has the highest priority and application threads on the application layer have
```
Comment (OlegHahm) [inline sys/net/gnrc/doc.txt]: I think you can omit one "application" here.

**Comment by cgundogan** [comment]
addressed comments of @OlegHahm

**Comment by authmillenon** [comment]
Since this isn't really an addition, but an enhancement, can you update the PR title and commit message please?

File: sys/net/gnrc/doc.txt [hunk 7]
```diff
@@ -0,0 +1,163 @@
+/**
+ * @defgroup    net_gnrc Generic (gnrc) network stack.
+ * @ingroup     net
+ * @brief       RIOT's modular default IP network stack.
+ *
+ * # About
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: Nitpicky, but I prefer the

```markdown
Section
======

Subsection
----------
```

syntax over the 

```markdown
# Section

## Subsection
```

syntax. I always think those are comments ^^

File: sys/net/gnrc/doc.txt [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+ *
+ * This module is currently the default network stack of RIOT and includes many components
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: `s/of/for/`

**Comment by authmillenon** [comment]
Slightly unrelated but can you `@ref net_gnrc "GNRC"` GNRC mentions in the `mainpage.md`, please?

File: sys/net/gnrc/doc.txt [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+ * ![GNRC Network Stack](riot-gnrc.svg)
+ * Each layer of the network stack runs in its own thread and each lower layer thread has a higher
+ * priority than any upper layer thread. In this regard, the thread of the MAC layer
+ * implementation has the highest priority and threads on the application layer have
+ * the lowest priority. The communication between threads is handled by the kernel's
+ * @ref core_msg functionality and by the @ref net_gnrc_netapi. Most of the times inter-process
+ * communication will take place between threads of neighboring layers for packets that traverse
+ * the network stack up or down.
+ *
+ * Due to the design of the @ref net_gnrc and the nature of inter-process communication,
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: `@ref net_gnrc "GNRC"` otherwise it looks weird in the compiled version

File: sys/net/gnrc/doc.txt [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+ * it is crucial for a new module that introduces a new thread to follow a certain programming
+ * construct if it desires to interact with other threads without blocking the system:
+ * Utilizing an `event loop`.
+ *
+ * Hence, a thread for the @ref net_gnrc will usually consist of four basic steps.
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: `@ref net_gnrc "GNRC"`

File: sys/net/gnrc/doc.txt [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+ * @defgroup    net_gnrc Generic (gnrc) network stack.
+ * @ingroup     net
+ * @brief       RIOT's modular default IP network stack.
+ *
+ * # About
+ *
+ * This module is currently the default network stack of RIOT and includes many components
+ * ranging from a simple MAC protocol (@ref net_gnrc_nomac) through a fully-featured IPv6
+ * implementation (@ref net_gnrc_ipv6) with 6LowPAN extensions (@ref net_gnrc_sixlowpan) to
+ * an UDP implementation (@ref net_gnrc_udp) and @ref net_gnrc_rpl.
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: Might be beneficial to use the `@ref <module> "text"` syntax here too, to get rid of all the parenthesis

File: sys/net/gnrc/doc.txt [hunk 12]
```diff
@@ -0,0 +1,172 @@
+/**
+ * @defgroup    net_gnrc Generic (gnrc) network stack.
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: Doesn't it make much more sense to put this whole thing in `sys/include/net/gnrc.h`

File: sys/net/gnrc/doc.txt [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+ * ranging from a simple MAC protocol (@ref net_gnrc_nomac) through a fully-featured IPv6
+ * implementation (@ref net_gnrc_ipv6) with 6LowPAN extensions (@ref net_gnrc_sixlowpan) to
+ * an UDP implementation (@ref net_gnrc_udp) and @ref net_gnrc_rpl.
+ *
+ * A list of all features contained in the @ref net_gnrc is available in the `Modules` section above.
+ *
+ * Integration into RIOT
+ * =====================
+ *
+ * From the application layer the @ref net_gnrc can be accessed through the @ref net_gnrc_conn,
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: I would just link to `net_conn` here, everything else is implementation detail.

File: sys/net/gnrc/doc.txt [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+ * while the interface to the @ref drivers_netdev_netdev2 is defined by the @ref net_gnrc_netdev2.
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: Unrelated, but both titles of these pages are quite a mouthful ^^

File: sys/net/gnrc/doc.txt [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+ * it is crucial for a new module that introduces a new thread to follow a certain programming
+ * construct if it desires to interact with other threads without blocking the system:
+ * Utilizing an `event loop`.
+ *
+ * Hence, a thread for @ref net_gnrc "GNRC" will usually consist of four basic steps.
+ * -# Initialize a message queue
+ * -# register for a @ref net_gnrc_nettype
+ * -# wait for a message
+ * -# react appropriately to a message and return to 3.
+ * @code{c}
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: Use

    ``` {.c}
    // ...
    ```

syntax, please

File: sys/net/gnrc/doc.txt [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+ * static msg_t _msg_q[Q_SZ];
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: I would prefer surrounding this whole code chunk with a function block ;-)

File: sys/net/gnrc/doc.txt [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+ * communication will take place between threads of neighboring layers for packets that traverse
+ * the network stack up or down.
+ *
+ * Due to the design of @ref net_gnrc "GNRC" and the nature of inter-process communication,
+ * it is crucial for a new module that introduces a new thread to follow a certain programming
+ * construct if it desires to interact with other threads without blocking the system:
+ * Utilizing an `event loop`.
+ *
+ * Hence, a thread for @ref net_gnrc "GNRC" will usually consist of four basic steps.
+ * -# Initialize a message queue
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: You can use [numbered lists](https://www.stack.nl/~dimitri/doxygen/manual/markdown.html#mddox_lists). This notation is historical.

File: sys/net/gnrc/doc.txt (hunk 15 above)
Comment (cgundogan) [inline sys/net/gnrc/doc.txt]: doesn't work

File: sys/net/gnrc/doc.txt [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+ *          case GNRC_NETAPI_MSG_TYPE_GET:
+ *             msg_reply(&msg, &reply);
+ *             break;
+ *         default:
+ *             break;
+ *     }
+ * }
+ * @endcode
+ * @note It is necessary to react with a reply message to the other thread for
+ *       @ref GNRC_NETAPI_MSG_TYPE_SET and @ref GNRC_NETAPI_MSG_TYPE_GET commands
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: Should we mention here too, that it is necessary to unregister `me_reg` again, if you leave the function context?

File: sys/net/gnrc/doc.txt [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+ * if (!gnrc_netapi_dispatch_send(GNRC_NETTYPE_UDP, 80, pkt)) {
+ *      puts("Error: no thread is interested");
+ *      gnrc_pktbuf_release(pkt);
+ *      return;
+ * }
+ * @endcode
+ * The @ref net_gnrc_pkt "pkt" will be send to all threads that registered for @ref GNRC_NETTYPE_UDP
+ * and the demux context `80`. Every registered thread will receive a @ref GNRC_NETAPI_MSG_TYPE_SND
+ * command and can access the @ref net_gnrc_pkt. If @ref gnrc_netapi_dispatch_send is replaced by
+ * @ref gnrc_netapi_dispatch_receive then threads will receive the @ref GNRC_NETAPI_MSG_TYPE_RCV
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: "then the thread will *send*"

File: sys/net/gnrc/doc.txt [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+ * instead, again with access to the @ref net_gnrc_pkt.
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: "*command* instead"

File: sys/net/gnrc/doc.txt [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+ * @code{c}
+ * if (!gnrc_netapi_dispatch_send(GNRC_NETTYPE_UDP, 80, pkt)) {
+ *      puts("Error: no thread is interested");
+ *      gnrc_pktbuf_release(pkt);
+ *      return;
+ * }
+ * @endcode
+ * The @ref net_gnrc_pkt "pkt" will be send to all threads that registered for @ref GNRC_NETTYPE_UDP
+ * and the demux context `80`. Every registered thread will receive a @ref GNRC_NETAPI_MSG_TYPE_SND
+ * command and can access the @ref net_gnrc_pkt. If @ref gnrc_netapi_dispatch_send is replaced by
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: I prefer to append `()` to function names to make unmistakeably clear that this is a function.

File: sys/net/gnrc/doc.txt [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+ * application's Makefile.
+ * - @code
+ * USEMODULE += gnrc_netif_default
+ * @endcode
+ * This will include necessary network drivers. If an IEEE802.15.4 device is present, then 6LoWPAN
+ * will also be included.
+ * - @code
+ * USEMODULE += auto_init_gnrc_netif
+ * @endcode
+ * This will initialize the network devices.
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: *auto-*initialize

File: sys/net/gnrc/doc.txt [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+ * - You may choose to build either as an IPv6 Node or as an IPv6 Router
+ *  -# IPv6 Node:
+ *  @code
```
Comment (authmillenon) [inline sys/net/gnrc/doc.txt]: Indentation is a bit off.

Comment (cgundogan) [inline sys/net/gnrc/doc.txt]: not really - it is meant to be nested

File: sys/net/gnrc/doc.txt (hunk 19 above)
Comment (cgundogan) [inline sys/net/gnrc/doc.txt]: classified as bullshit after discussing offline about this

File: sys/net/gnrc/doc.txt (hunk 16 above)
Comment (cgundogan) [inline sys/net/gnrc/doc.txt]: I am not so sure about this - this would imply a working example. However, this code snipped depicted here is just a sketch and not ready to compile.

File: sys/net/gnrc/doc.txt (hunk 14 above)
Comment (cgundogan) [inline sys/net/gnrc/doc.txt]: Any proposals?

**Comment by cgundogan** [comment]
should I squash?

File: sys/include/net/gnrc.h [hunk 24]
```diff
@@ -7,9 +7,201 @@
...
+ * Utilizing an `event loop`.
+ *
+ * Hence, a thread for @ref net_gnrc "GNRC" will usually consist of four basic steps.
+ * 1. Initialize a message queue
+ * 2. register for a @ref net_gnrc_nettype
+ * 3. wait for a message
+ * 4. react appropriately to a message and return to 3.
+ *
+ * @code{c}
+ * static msg_t _msg_q[Q_SZ];
```
Comment (authmillenon) [inline sys/include/net/gnrc.h]: Function context still missing.

File: sys/include/net/gnrc.h [hunk 25]
```diff
@@ -7,9 +7,201 @@
...
+ * module. In order to receive a @ref net_gnrc_pkt of a specific type, it is necessary to register
+ * for the appropriate @ref net_gnrc_nettype first. Your thread will then be able to receive
+ * certain commands defined in the @ref net_gnrc_netapi module
+ * (e.g. @ref net_gnrc_netapi::GNRC_NETAPI_MSG_TYPE_RCV) for all @ref net_gnrc_pkt "Packets"
+ * that your thread registered for.
+ *
+ * The following example will sketch how to receive incoming and outgoing UDP traffic on port 80.
+ *
+ * @code{c}
+ * static msg_t _msg_q[Q_SZ];
```
Comment (authmillenon) [inline sys/include/net/gnrc.h]: Function context still missing

**Comment by cgundogan** [comment]
@authmillenon addressed comment

**Comment by authmillenon** [comment]
Args sorry accidentally commented in commit >.<

**Comment by cgundogan** [comment]
which line breaks?

**Comment by authmillenon** [comment]
```c
int main(void) {
```

vs

```c
int main() 
{
```

**Comment by cgundogan** [comment]
addressed @authmillenon's comment

**Comment by cgundogan** [comment]
@authmillenon should I squash?

**Comment by authmillenon** [comment]
Please consider https://github.com/cgundogan/RIOT/pull/8 before squashing.

**Comment by cgundogan** [comment]
@authmillenon I merged your commit but it breaks the doc. Do you know how to fix this?

**Comment by authmillenon** [comment]
It seems to be a parsing issue with 1.8.10. Let's hope the 1.8.9.1 on the doc server does not has this issue (since with 1.8.6 it worked fine)

**Comment by cgundogan** [comment]
squashed

**Comment by authmillenon** [comment]
ACK and go, when travis is happy.

File: doc/doxygen/src/riot-gnrc.svg [hunk 26]
```diff
@@ -0,0 +1,588 @@
...
+          <text id="gnrc-conn-label" class="modbox-text"
+                x="190" y="32"><tspan>gnrc_conn</tspan></text>
+        </g>
+        <g id="netapi-conn-udp">
+          <path id="netapi-conn-udp-arrow" d="m 97.5,55 0,20" class="outer-stroke-arrow" />
+          <text id="netapi-conn-udp-label" class="outer-text-netapi"
+                x="104" y="71"><tspan>netapi</tspan></text>
+        </g>
+        
+         <g id="netapi-conn-ipv6">
```
Comment (authmillenon) [inline doc/doxygen/src/riot-gnrc.svg]: Whitespace error in 512 and indentation error

File: doc/doxygen/src/riot-gnrc.svg [hunk 27]
```diff
@@ -0,0 +1,588 @@
...
+          <path id="netapi-conn-udp-arrow" d="m 190.5,55 0,100" class="outer-stroke-arrow" />
+          <text id="netapi-conn-udp-label" class="outer-text-netapi"
+                x="195" y="71"><tspan>netapi</tspan></text>
+        </g>
+
```
Comment (authmillenon) [inline doc/doxygen/src/riot-gnrc.svg]: Whitespace error (this line can be removed IMHO)

**Comment by authmillenon** [comment]
Regarding https://github.com/RIOT-OS/RIOT/pull/4415#issuecomment-162867979: reported it to doxygen https://bugzilla.gnome.org/show_bug.cgi?id=759177

**Comment by cgundogan** [comment]
@authmillenon: addressed your comments and squashed

**Comment by authmillenon** [comment]
Then gogogo. ;-)

**Comment by OlegHahm** [comment]
Cool! Thanks for the effort!

**Comment by OlegHahm** [comment]
Hm, maybe it's stupid, but I think this should be added to the release branch, too.

**Comment by authmillenon** [comment]
Doxygen 1.8.9.1 does not have a problem with the fenced code either :)
